File: SGUEES-API/sguees.api/Options/SelectionHiring/SC_INDUCCION/SC_INDUCCIONService.cs
// Qué hace: aplica las reglas de negocio del catálogo inducción antes de llamar al repositorio.
using System.Collections.Generic;
using System.Threading.Tasks;
using eFramework.Core;
using SGUEES.Models;
using SGUEES.Repositories;

namespace SGUEES.Services
{
    // Qué hace: valida los datos de inducción y coordina su persistencia con el repositorio.
    public class SC_INDUCCIONService : ISC_INDUCCIONService
    {
        private readonly ISC_INDUCCIONRepository _repo;

        public SC_INDUCCIONService(ISC_INDUCCIONRepository repo)
        {
            _repo = repo;
        }

        // Qué hace: entrega las inducciones activas disponibles para el descriptor.
        // Cómo: llama a GetCatalogoDescriptorAsync del repositorio y arma el CResult con el listado.
        public async Task<CResult> GetCatalogoDescriptorAsync(SC_INDUCCIONParam xWhere)
        {
            var rows = await _repo.GetCatalogoDescriptorAsync(xWhere.CORR_EMPRESA);
            return new CResult
            {
                Data = rows,
                Result = true,
                RowsAffected = rows.Count,
                CodeHelper = 0,
                ErrorCode = 0,
                ErrorMessage = "",
                ErrorSource = "",
            };
        }

        // Qué hace: lista las inducciones según los filtros recibidos.
        // Cómo: llama a GetAllAsync del repositorio con los parámetros armados en BuildParameters.
        public async Task<CResult> GetAllAsync(SC_INDUCCIONParam xWhere)
        {
            return await _repo.GetAllAsync(BuildParameters(xWhere));
        }

        // Qué hace: obtiene una inducción por su correlativo.
        // Cómo: llama a GetAsync del repositorio con CORR_EMPRESA y CORR_INDUCCION.
        public async Task<CResult> GetAsync(SC_INDUCCIONParam xWhere)
        {
            var p = new List<CParameter>
            {
                new CParameter() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
                new CParameter() { ParameterName = "CORR_INDUCCION", Value = xWhere.CORR_INDUCCION, DbType = System.Data.DbType.Int32 },
            };

            return await _repo.GetAsync(p);
        }

        // Qué hace: crea una inducción nueva.
        // Cómo: valida empresa de sesión y datos con Validate, normaliza el nombre y llama a CreateAsync del repositorio.
        public async Task<CResult> CreateAsync(SC_INDUCCIONTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
            if (empresaError != null)
            {
                return empresaError;
            }

            var validation = Validate(Data);
            if (validation != null)
            {
                return validation;
            }

            NormalizeData(Data);
            return await _repo.CreateAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: actualiza una inducción existente.
        // Cómo: valida empresa, datos y llave; normaliza el nombre y llama a UpdateAsync del repositorio.
        public async Task<CResult> UpdateAsync(SC_INDUCCIONTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
            if (empresaError != null)
            {
                return empresaError;
            }

            var validation = Validate(Data);
            if (validation != null)
            {
                return validation;
            }

            if (Data.CORR_INDUCCION <= 0)
            {
                return ValidationError("No se pudo identificar la induccion a actualizar.");
            }

            NormalizeData(Data);
            return await _repo.UpdateAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: elimina una inducción.
        // Cómo: valida la empresa de sesión y llama a DeleteAsync del repositorio.
        public async Task<CResult> DeleteAsync(SC_INDUCCIONTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
            if (empresaError != null)
            {
                return empresaError;
            }

            return await _repo.DeleteAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: cambia el estado activo/inactivo de una inducción.
        // Cómo: valida empresa y llave; llama a ActivarInactivarAsync del repositorio.
        public async Task<CResult> ActivarInactivarAsync(SC_INDUCCIONTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
            if (empresaError != null)
            {
                return empresaError;
            }

            if (Data.CORR_INDUCCION <= 0)
            {
                return ValidationError("No se pudo identificar la induccion a actualizar.");
            }

            return await _repo.ActivarInactivarAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: arma el parámetro CORR_EMPRESA para filtrar en el repositorio.
        private static List<CParameter> BuildParameters(SC_INDUCCIONParam xWhere)
        {
            return new List<CParameter>
            {
                new CParameter() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
            };
        }

        // Qué hace: normaliza los datos antes de guardar.
        // Cómo: recorta espacios del nombre y fija ESTADO_INDUCCION en true si viene vacío.
        private static void NormalizeData(SC_INDUCCIONTable Data)
        {
            Data.NOMBRE_INDUCCION = Data.NOMBRE_INDUCCION?.Trim();
            Data.ESTADO_INDUCCION ??= true;
        }

        // Qué hace: valida los datos de la inducción.
        // Cómo: revisa que existan datos, que el nombre no esté vacío ni supere 200 caracteres, y que las semanas sean mayores a 0.
        private static CResult Validate(SC_INDUCCIONTable Data)
        {
            if (Data == null)
            {
                return ValidationError("No se recibieron datos de induccion.");
            }

            if (string.IsNullOrWhiteSpace(Data.NOMBRE_INDUCCION))
            {
                return ValidationError("Debe ingresar el nombre de induccion.");
            }

            if (Data.NOMBRE_INDUCCION.Trim().Length > 200)
            {
                return ValidationError("El nombre de induccion no puede superar 200 caracteres.");
            }

            if (Data.SEMANAS_INDUCCION <= 0)
            {
                return ValidationError("Debe ingresar semanas de induccion mayores a 0.");
            }

            return null;
        }

        // Qué hace: valida que exista empresa en la sesión.
        // Cómo: si CORR_EMPRESA es mayor a 0 permite continuar; si no, devuelve un CResult de error.
        private static CResult ValidateEmpresaSesion(int corrEmpresa)
        {
            if (corrEmpresa > 0)
            {
                return null;
            }

            return new CResult
            {
                Data = null,
                Result = false,
                CodeHelper = 0,
                ErrorCode = 4100,
                ErrorMessage = "No se pudo guardar la inducción porque su usuario no tiene una empresa asignada. Solicite que le configuren una empresa por defecto en el sistema.",
                ErrorSource = "[SC_INDUCCIONService]",
                RowsAffected = 0
            };
        }

        // Qué hace: construye un CResult de error de validación con el mensaje recibido.
        private static CResult ValidationError(string message)
        {
            return new CResult
            {
                Data = null,
                Result = false,
                CodeHelper = 0,
                ErrorCode = -1,
                ErrorMessage = message,
                ErrorSource = "[SC_INDUCCIONService]",
                RowsAffected = 0
            };
        }
    }
}

File: SGUEES-SPA/src/app/pages/SelectionHiring/sc-induccion/sc-induccion.service.ts
// Qué hace: agrupa las reglas de negocio del catálogo Inducción.
// Cómo: valida los datos y llama al repositorio para el CRUD y el cambio de estado; define columnas y campos del formulario.
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { IParam } from 'src/app/FxAPI/IParam';
import { IResult } from 'src/app/FxAPI/IResult';
import { NotifyType } from 'src/app/shared/models/NotifyType';
import { buildAuditGridColumns } from 'src/app/shared/mtto/mtto-grid.helpers';
import { createEstadoColumnConfig, ESTADO_ACTIVO_INACTIVO_LABELS } from 'src/app/shared/utils/remote-grid-filter.util';
import {
	ScInduccion,
	UNIDAD_TIEMPO_MESES,
	UNIDAD_TIEMPO_SEMANAS,
} from './models/sc-induccion';
import { ScInduccionRepository } from './sc-induccion.repository';

const ESTADO_FIELD = 'ESTADO_INDUCCION';
const TIEMPO_INDUCCION_MIN = 1;
const TIEMPO_INDUCCION_MAX = 100;

type ScInduccionFormOptions = {
	unidadesTiempo?: Array<{ Key: any; Value: string }>;
};

@Injectable({ providedIn: 'root' })
// Qué hace: valida los datos de inducción y coordina el CRUD con el repositorio.
export class ScInduccionService {
	constructor(private repo: ScInduccionRepository) {}

	// Qué hace: valida los datos de la inducción antes de guardar.
	// Cómo: revisa nombre (máx. 100), tiempo entre 1 y 100 (si está fuera muestra aviso y no guarda) y unidad Semanas/Meses.
	esValido(model: ScInduccion, msg: Function): boolean {
		if (!model.NOMBRE_INDUCCION || model.NOMBRE_INDUCCION.trim() === '') {
			msg('Debe ingresar el nombre de induccion.', NotifyType.Warning);
			return false;
		}

		if (model.NOMBRE_INDUCCION.trim().length > 100) {
			msg('El nombre de induccion no puede superar 100 caracteres.', NotifyType.Warning);
			return false;
		}

		const tiempo = Number(model.TIEMPO_INDUCCION);
		if (!Number.isFinite(tiempo) || tiempo < TIEMPO_INDUCCION_MIN || tiempo > TIEMPO_INDUCCION_MAX) {
			msg('El tiempo de induccion debe estar entre 1 y 100.', NotifyType.Warning);
			return false;
		}

		const unidad = (model.UNIDAD_TIEMPO ?? '').trim();
		if (unidad !== UNIDAD_TIEMPO_SEMANAS && unidad !== UNIDAD_TIEMPO_MESES) {
			msg('Debe seleccionar la unidad de tiempo (Semanas o Meses).', NotifyType.Warning);
			return false;
		}

		return true;
	}

	// Qué hace: lista las inducciones según los filtros recibidos.
	// Cómo: llama a getAll del repositorio con los parámetros armados en buildWhere.
	getAll(param: any): Observable<IResult> {
		return this.repo.getAll(this.buildWhere(param));
	}

	// Qué hace: obtiene una inducción por su correlativo.
	// Cómo: llama a get del repositorio con CORR_INDUCCION como filtro.
	get(param: any): Observable<IResult> {
		return this.repo.get([{ Parameter: 'CORR_INDUCCION', Value: param.CORR_INDUCCION }]);
	}

	// Qué hace: crea una inducción nueva.
	// Cómo: llama a create del repositorio con el modelo recibido.
	insert(model: any): Observable<IResult> {
		return this.repo.create(model);
	}

	// Qué hace: actualiza una inducción existente.
	// Cómo: llama a update del repositorio con el modelo y CORR_INDUCCION como llave.
	update(model: any): Observable<IResult> {
		return this.repo.update(model, [{ Parameter: 'CORR_INDUCCION', Value: model.CORR_INDUCCION }]);
	}

	// Qué hace: elimina una inducción.
	// Cómo: llama a delete del repositorio con CORR_INDUCCION como filtro.
	delete(model: any): Observable<IResult> {
		return this.repo.delete([{ Parameter: 'CORR_INDUCCION', Value: model.CORR_INDUCCION }]);
	}

	// Qué hace: cambia el estado activo/inactivo de una inducción.
	// Cómo: llama a activarInactivar del repositorio con CORR_INDUCCION como filtro.
	activarInactivar(model: any): Observable<IResult> {
		return this.repo.activarInactivar(model, [{ Parameter: 'CORR_INDUCCION', Value: model.CORR_INDUCCION }]);
	}

	// Qué hace: define las columnas de la grilla de mantenimiento.
	// Cómo: arma columnas de correlativo, nombre, tiempo, unidad, estado y auditoría.
	getColumns(): any {
		return [
			{
				dataField: 'CORR_INDUCCION',
				caption: 'Corr.',
				width: 90,
				dataType: 'number',
				filterOperations: ['=', '<', '>', '<=', '>='],
			},
			{ dataField: 'NOMBRE_INDUCCION', caption: 'Induccion', width: 280 },
			{
				dataField: 'TIEMPO_INDUCCION',
				caption: 'Tiempo',
				width: 100,
				dataType: 'number',
				filterOperations: ['=', '<', '>', '<=', '>='],
			},
			{ dataField: 'UNIDAD_TIEMPO', caption: 'Unidad', width: 120 },
			createEstadoColumnConfig(ESTADO_FIELD, ESTADO_ACTIVO_INACTIVO_LABELS),
			...buildAuditGridColumns({ withDateTimeFilter: true }),
		];
	}

	// Qué hace: define el resumen (contador) de la grilla.
	// Cómo: cuenta filas sobre CORR_INDUCCION con formato Cant: {0}.
	getSummary(): any {
		return {
			totalItems: [
				{
					column: 'CORR_INDUCCION',
					summaryType: 'count',
					valueFormat: '#,##0',
					displayFormat: 'Cant: {0}',
				},
			],
		};
	}

	// Qué hace: define los campos y las reglas de validación del formulario.
	// Cómo: usa unidadesTiempo (getLookUp SC_LISTA) como dataSource del SelectBox de UNIDAD_TIEMPO;
	//       el tiempo se valida entre 1 y 100 al guardar (sin recortar el valor al escribir).
	getItems(options?: ScInduccionFormOptions): any {
		const unidadesTiempo = options?.unidadesTiempo ?? [];

		return [
			{ dataField: 'CORR_INDUCCION', label: { text: 'Corr.' }, colSpan: 1, editorOptions: { readOnly: true } },
			{
				dataField: 'NOMBRE_INDUCCION',
				label: { text: 'Nombre induccion' },
				colSpan: 5,
				editorOptions: { placeholder: 'Nombre induccion...', showClearButton: true, maxLength: 100 },
				validationRules: [{ type: 'required', message: 'Este campo es obligatorio' }],
			},
			{
				dataField: 'TIEMPO_INDUCCION',
				label: { text: 'Tiempo' },
				editorType: 'dxNumberBox',
				colSpan: 2,
				editorOptions: {
					step: 1,
					showSpinButtons: true,
					format: '#0',
				},
				validationRules: [
					{ type: 'required', message: 'Este campo es obligatorio' },
					{
						type: 'range',
						min: TIEMPO_INDUCCION_MIN,
						max: TIEMPO_INDUCCION_MAX,
						message: 'El tiempo debe estar entre 1 y 100',
					},
				],
			},
			{
				dataField: 'UNIDAD_TIEMPO',
				label: { text: 'Unidad de tiempo' },
				editorType: 'dxSelectBox',
				colSpan: 2,
				editorOptions: {
					dataSource: unidadesTiempo,
					valueExpr: 'Key',
					displayExpr: 'Value',
					searchEnabled: false,
					showClearButton: false,
				},
				validationRules: [{ type: 'required', message: 'Este campo es obligatorio' }],
			},
			{ dataField: 'ESTADO_INDUCCION', label: { text: 'Activo' }, editorType: 'dxCheckBox', colSpan: 2 },
		];
	}

	// Qué hace: arma los filtros de consulta a partir de los parámetros recibidos.
	// Cómo: agrega CORR_INDUCCION a xWhere solo cuando viene informado.
	private buildWhere(param: any): IParam[] {
		const xWhere: IParam[] = [];

		if (param.CORR_INDUCCION) {
			xWhere.push({ Parameter: 'CORR_INDUCCION', Value: param.CORR_INDUCCION });
		}

		return xWhere;
	}
}
